sdk/cosmos/src/authorization_policy.rs

```diff
@@ -0,0 +1,320 @@
+        Self {
+            authorization_token,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) struct CosmosContext {
+    pub(crate) resource_type: ResourceType,
+}
+
+impl From<ResourceType> for CosmosContext {
+    fn from(resource_type: ResourceType) -> Self {
+        Self { resource_type }
+    }
+}
+
+#[async_trait::async_trait]
+impl Policy<CosmosContext> for AuthorizationPolicy {
+    async fn send(
+        &self,
+        ctx: &mut PipelineContext<CosmosContext>,
+        request: &mut Request,
+        next: &[Arc<dyn Policy<CosmosContext>>],
+    ) -> PolicyResult<Response> {
+        trace!("called AuthorizationPolicy::send. self == {:#?}", self);
+
+        if next.is_empty() {
+            return Err(Box::new(azure_core::PipelineError::InvalidTailPolicy(
+                "Authorization policies cannot be the last policy of a pipeline".to_owned(),
+            )));
+        }
+
+        let time = chrono::Utc::now().format(TIME_FORMAT).to_string();
+
+        let uri_path = &request.uri().path_and_query().unwrap().to_string()[1..];
+        trace!("uri_path used by AuthorizationPolicy == {:#?}", uri_path);
+
+        let auth = {
+            let resource_link = generate_resource_link(&uri_path);
+            debug!("resource_link == {}", resource_link);
+            generate_authorization(
+                &self.authorization_token,
+                &request.method(),
+                &ctx.get_bag().resource_type,
+                resource_link,
+                &time,
+            )
+        };
+
+        trace!(
+            "AuthorizationPolicy calculated authorization == {}: {}",
+            AUTHORIZATION,
+            &auth
+        );
+
+        request
+            .headers_mut()
+            .append(HEADER_DATE, HeaderValue::from_str(&time)?);
+        request
+            .headers_mut()
+            .append(HEADER_VERSION, HeaderValue::from_static(AZURE_VERSION));
+        request
+            .headers_mut()
+            .append(AUTHORIZATION, HeaderValue::from_str(&auth)?);
+
+        trace!("\n\nrequest =={:?}", request);
+
+        // now next[0] is safe (will not panic) because we checked
+        // at the beginning of the function.
+        next[0].send(ctx, request, &next[1..]).await
+    }
+}
+
+// TODO: will become private as soon as cosmos_client will be migrated
+// to pipeline arch.
+pub(crate) fn generate_resource_link(u: &str) -> &str {
+    static ENDING_STRINGS: &[&str] = &[
+        "dbs",
+        "colls",
+        "docs",
+        "sprocs",
+        "users",
+        "permissions",
+        "attachments",
+        "pkranges",
+        "udfs",
+        "triggers",
+    ];
+
+    // store the element only if it does not end with dbs, colls or docs
+    let p = u;
+    let len = p.len();
+    for str_to_match in ENDING_STRINGS {
+        let end_len = str_to_match.len();
+
+        if end_len <= len {
+            let end_offset = len - end_len;
```

Comment:
Underflow is possible here which we might want to protect against.

Reply:
I've completely rewritten the function in a more functional manner. It should be safe now.